### ISSUE TITLE
Unauthorized to download public package?

### ISSUE THREAD

**Issue by jakirkham** [conda/conda#3004]
So, this one is pretty confusing. This is happening with `conda` 4.1.7 (hot off the press). I recently restarted the CI of a previously passing conda-forge PR ( https://github.com/conda-forge/numpy-feedstock/pull/17 ) that sat for a few days to see if it was merge ready. While the PR consists of little more that a version bump it failed. While that was surprising, what was really surprising was the error message presented.

```
Error: HTTPError: 401 Client Error: UNAUTHORIZED for url: https://conda.anaconda.org/t/None/conda-forge/linux-64/ca-certificates-2016.2.28-1.tar.bz2: https://conda.anaconda.org/t/None/conda-forge/linux-64/ca-certificates-2016.2.28-1.tar.bz2
```

For more details, please see this [log]( https://circleci.com/gh/conda-forge/numpy-feedstock/50 ).

**Comment by mwcraig** [comment]
Ditto, just started seeing this. Sample log [here](https://travis-ci.org/BruceSherwood/vpython-jupyter/jobs/143173502#L308)

**Comment by jakirkham** [comment]
FWIW I think this is an easy fix. Somehow we are inserting a token here when there is no token. This URL doesn't work

```
https://conda.anaconda.org/t/None/conda-forge/linux-64/ca-certificates-2016.2.28-1.tar.bz2
```

However, this URL does.

```
https://conda.anaconda.org/conda-forge/linux-64/ca-certificates-2016.2.28-1.tar.bz2
```

Probably a simple case of `%` or `.format` gone wrong. A `None` check before proceeding should trivially fix this.

**Comment by jakirkham** [comment]
I'm going to point my grubby little finger at this [line]( https://github.com/conda/conda/compare/4.1.6...4.1.7#diff-3b5a20f4aa6ecedbf4250f6d24282af6R244 ) (and the one after). Probably a simple matter of adding `binstar_client.token` to the `if`. 😄

**Comment by jakirkham** [comment]
Also cc @ocefpaf @jjhelmus for awareness before going down the MNT PR road.

**Comment by mcg1969** [comment]
Yes, that's exactly the problem. CRAP.

**Comment by mcg1969** [comment]
Thinking we should go ahead and pull 4.1.7.

**Comment by jakirkham** [comment]
Given the severity of the problem, I would agree with you.

**Comment by mcg1969** [comment]
Or we can get the Anaconda Cloud team to create a login token called 'None' :-)

**Comment by mcg1969** [comment]
Pulled.

**Comment by mwcraig** [comment]
Looking forward to conda canary! 😀

**Comment by mcg1969** [comment]
Indeed! But I do hope that as many of you as possible participate in this. I'm certainly frustrated with my own performance here. If you look at the integration testing we've been adding we're truly making an effort to address every fix with the right tests. But we're just not catching all of the right cases internally. We will get there but it is a slog. Canary is _definitely_ the right way to go.

**Comment by jakirkham** [comment]
@msarahan and I were discussing the potential of flinging a bunch of PRs at conda-forge feedstocks that test out dev releases of `conda-build`. Maybe something similar could be explored for `conda`. I think part of the struggle that is occurring is simply we need a broader array of test cases and we need the results sooner. This might help with that problem.

**Comment by mwcraig** [comment]
I've been assuming that some of the conda-forge sprint time will be devoted to writing conda and conda-build tests. They are great tools!

**Comment by mcg1969** [comment]
> I think part of the struggle that is occurring is simply we need a broader array of test cases

This is the nut of the issue. When we've tried a traditional beta approach, we don't get a lot of participation. I understand that---people just want to Get Things Done. At the same time, our testing just doesn't hit all of the corner cases we see in the wild. (To be fair, this _particular_ bug can simply be chalked up to my lack of creativity.)

**Comment by mwcraig** [comment]
@mcg1969 -- I'm seeing this error again on travis, which still upgrades to 4.1.7: https://travis-ci.org/OpenAstronomy/conda-channel/jobs/143185136#L210

Oddly, a build I restarted a few hours ago after you pulled 4.1.7 succeeded because travis only updated to 4.1.6.

**Comment by jakirkham** [comment]
No chance you guys are caching the packages, @mwcraig?

**Comment by mwcraig** [comment]
Not intentionally, no, @jakirkham. I'm really confused, because I've never set up caching, `conda search conda` on my mac does **not** find 4.1.7, and a job I started in another repo on travis also does not find 4.1.7.

**Comment by mwcraig** [comment]
@mcg1969 - Never mind, fixed by closing and reopening the pull request. No idea why restarting the build didn't work...

**Comment by mwcraig** [comment]
One suggestion regarding increasing the range of test cases: if it were easy to add an entry to build matrix on travis that tested building against pre-releases, or even master, of `conda` and `conda-build`, I'd be happy to do that case on all of the builds I'm maintaining.

**Comment by kalefranz** [comment]
closed with #3008
